src/canvasObject.js

```diff
@@ -1,159 +1,128 @@
+      },
+      self,
+      dispatcher
+    );
+  });
+
   this.label = config.canvas.label;
-  this.tileSourceUrl = config.canvas.images[0].resource.service['@id'] + '/info.json';
   this.viewingHint = config.canvas.viewingHint;
 
   this.dispatcher = dispatcher;
 };
 
 CanvasObject.prototype = {
-  openTileSource: function(viewer) {
-    var self = this;
+  openTileSource: function(viewer, imageIndex) {
+    this.dispatcher.emit('detail-tile-source-opened', { 'detail': this.id });
+    var thumbnail = this.thumbnail;
+    var image = this.images[imageIndex];
+
+    var onTileDrawn = function(event) {
+      if(event.detail === image.tileSource) {
+        image.hide(true);
+        image.fade(1);
+
+        if(thumbnail){
+          thumbnail.destroy(viewer);
+        }
+      }
+    };
 
-    // We've already loaded this tilesource instead of the thumbnail
-    if(this.fullyOpened) {
-      return;
-    }
-
-    // otherwise, continue loading the tileSource.
-    this.dispatcher.emit('detail-tile-source-requested', { 'detail': this.id });
-    viewer.addTiledImage({
-      x: this.x,
-      y: this.y,
-      width: this.width,
-      tileSource: this.tileSourceUrl,
-      opacity: this.opacity,
-      clip: this.clipRegion,
-      index: 0, // Add the new image below the stand-in.
-
-      success: function(event) {
-        var main = event.item;
-        self.fullyOpened = true;
-
-        var tileDrawnHandler = function(event) {
-          if (event.tiledImage === main) {
-            var previousImageObj = self._mainImageObj;
-
-            viewer.removeHandler('tile-drawn', tileDrawnHandler);
-            main.setOpacity(0, true);
-            self._fade(main, 1);
-            self._setMainImage(main);
-
-            if(previousImageObj){
-              viewer.world.removeItem(previousImageObj);
-            }
-            self.dispatcher.emit('detail-tile-source-opened', { 'detail': self.id });
-          }
-        };
-        viewer.addHandler('tile-drawn', tileDrawnHandler);
-      },
+    this.dispatcher.once('image-resource-tile-source-opened', onTileDrawn);
+    image.openTileSource(viewer);
+  },
 
-      error: function(event) {
-        var errorInfo = {
-          id: self.id,
-          message: event.message,
-          source: event.source
-        };
-        self.dispatcher.emit('detail-tile-source-failed', {'detail': errorInfo});
-      }
-    });
+  openMainTileSource: function(viewer) {
+    this.openTileSource(viewer, 0);
   },
 
   openThumbnail: function(viewer) {
+    this.dispatcher.emit('detail-thumbnail-opened', { 'detail': this.id });
     var self = this;
-    viewer.addTiledImage({
-      x: this.x,
-      y: this.y,
-      width: this.width,
-      tileSource: this.placeholder,
-      opacity: this.opacity,
-      clip: this.clipRegion,
-      success: function(event) {
-        self._setMainImage(event.item);
-      }
-    })
+
+    this.thumbnail = new ImageResource(
```

Comment:
The ImageResource constructor now has enough arguments it's probably worth changing over to a single object argument so we can get named params (like the config of CanvasObject).